java/src/main/java/ai/rapids/cudf/JCudfSerialization.java

```diff
@@ -621,52 +772,69 @@ public static long getSerializedSizeInBytes(HostColumnVector[] columns, long row
+  /** Build a list of column offset descriptors using a pre-order traversal of the columns */
+  static ArrayDeque<ColumnOffsets> buildIndex(SerializedTableHeader header,
+                                              HostMemoryBuffer buffer) {
+    int numTopColumns = header.getNumColumns();
+    ArrayDeque<ColumnOffsets> offsetsList = new ArrayDeque<>();
     long bufferOffset = 0;
-    DType[] dataTypes = header.types;
-    int numColumns = dataTypes.length;
-    long[] nullCounts = header.nullCounts;
-    long numRows = header.getNumRows();
-    ColumnOffsets[] ret = new ColumnOffsets[numColumns];
-    for (int column = 0; column < numColumns; column++) {
-      DType type = dataTypes[column];
-      long nullCount = nullCounts[column];
-
-      long validity = 0;
-      long validityLen = 0;
-      long offsets = 0;
-      long offsetsLen = 0;
-      long data = 0;
-      long dataLen = 0;
-      if (nullCount > 0) {
-        validityLen = padFor64byteAlignment(BitVectorHelper.getValidityLengthInBytes(numRows));
-        validity = bufferOffset;
-        bufferOffset += validityLen;
-      }
-
-      if (type == DType.STRING) {
-        if (numRows > 0) {
-          offsetsLen = (numRows + 1) * 4;
-          offsets = bufferOffset;
-          int startStringOffset = buffer.getInt(bufferOffset);
-          int endStringOffset = buffer.getInt(bufferOffset + (numRows * 4));
-          bufferOffset += padFor64byteAlignment(offsetsLen);
+    for (int i = 0; i < numTopColumns; i++) {
+      SerializedColumnHeader column = header.getColumnHeader(i);
+      bufferOffset = buildIndex(column, buffer, offsetsList, bufferOffset);
+    }
+    return offsetsList;
+  }
 
-          dataLen = endStringOffset - startStringOffset;
+  /**
+   * Append a list of column offset descriptors using a pre-order traversal of the column
+   * @param column column offset descriptors will be built for this column and its child columns
+   * @param buffer host buffer backing the column data
+   * @param offsetsList list where column offset descriptors will be appended during traversal
+   * @param bufferOffset offset in the host buffer where the column data begins
+   * @return buffer offset at the end of this column's data including all child columns
+   */
+  private static long buildIndex(SerializedColumnHeader column, HostMemoryBuffer buffer,
+                                 ArrayDeque<ColumnOffsets> offsetsList, long bufferOffset) {
+    long validity = 0;
+    long offsets = 0;
+    long data = 0;
+    long dataLen = 0;
+    long rowCount = column.getRowCount();
+    if (column.getNullCount() > 0) {
+      long validityLen = padFor64byteAlignment(BitVectorHelper.getValidityLengthInBytes(rowCount));
+      validity = bufferOffset;
+      bufferOffset += validityLen;
+    }
+
+    DType dtype = column.getType();
+    if (dtype.hasOffsets()) {
+      if (rowCount > 0) {
+        long offsetsLen = (rowCount + 1) * 4;
```

Comment:
nit: Should we make this `Integer.BYTES` too? There are several places that we have 4 hard coded.

Reply:
I agree it's inconsistent to switch between the two.  Shall I update all the `* 4` instances?  There are many more than just this one.

Reply:
I think it would be best so it is consistent, but it is minor.

Reply:
OK, I've updated places to use `Integer.BYTES` when dealing with offsets.